File: solutions/day_007/dart/valid_sudoku.dart
import 'dart:collection';

class Solution {
  bool isValidSudoku(List<List<String>> board) {
    // create a hashset to store the numbers in the row, column, and block
    HashSet<String> hashSet = HashSet();
    // loop through the board
    for (int row = 0; row < 9; row++) {
      // loop through the columns
      for (int column = 0; column < 9; column++) {
        // get the number at the current position
        String number = board[row][column];
        // if the number is not a period, add it to the hashset
        if (number != '.') {
          // if the number is already in the hashset, return false
          if (!hashSet.add(number + " in row " + String.fromCharCode(row)) ||
              !hashSet
                  .add(number + " in column " + String.fromCharCode(column)) ||
              !hashSet.add(
                number +
                    " in block " +
                    String.fromCharCode(row ~/ 3) +
                    "," +
                    String.fromCharCode(column ~/ 3),
              )) return false;
        }
      }
    }
    return true;
  }
}

File: solutions/day_007/java/valid_sudoku.java
package solutions.day_007.java;

import java.util.HashSet;

class Solution {
    public boolean isValidSudoku(char[][] board) {
        // 1. Create a set to store the numbers
        HashSet<String> hs = new HashSet<>();
        // 2. Check if the number is valid
        for (int i = 0; i < board.length; i++) {
            for (int j = 0; j < board[0].length; j++) {
                // 3. Add the number to the set
                if (board[i][j] != '.') {
                    String row = "row" + i + board[i][j];
                    String col = "col" + j + board[i][j];
                    int boxnum = ((i / 3) * 3) + (j / 3);
                    String box = "box" + boxnum + board[i][j];
                    if (hs.contains(row) || hs.contains(col) || hs.contains(box))
                        return false;
                    hs.add(row);
                    hs.add(col);
                    hs.add(box);
                }

            }
        }
        return true;
    }

}